File: full_stack_node/backend/src/routes/taint.routes.ts
/**
 * Taint Flow Routes - Express Backend
 *
 * These routes demonstrate cross-boundary taint flows from React frontend.
 * Each route has a clear Source → Sink path for taint analysis testing.
 *
 * TAINT FLOWS:
 * 1. /reports/generate - Command Injection (filename → exec)
 * 2. /files/:fileId/download - Path Traversal (fileId → fs.readFile)
 * 3. /webhooks/register - SSRF (url → axios.get)
 * 4. /users/advanced-search - NoSQL Injection (filter → MongoDB query)
 */

import { Router, Request, Response } from 'express';
import { exec } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import axios from 'axios';
import { sequelize } from '../config/database';
import { QueryTypes } from 'sequelize';

const router = Router();

/**
 * TAINT FLOW #1: Command Injection
 * Source: req.body.filename (from React ReportGenerator)
 * Sink: child_process.exec()
 * Path: HTTP POST body → exec()
 */
router.post('/reports/generate', async (req: Request, res: Response) => {
  const { filename, format, filters } = req.body;

  // TAINT: User controlled filename flows directly to shell command
  // Attacker payload: "; rm -rf /" or "$(cat /etc/passwd)"
  const command = `generate-report --output /tmp/reports/${filename}.${format}`;

  exec(command, (error, stdout, stderr) => {
    if (error) {
      // Leaking error details
      return res.status(500).json({
        success: false,
        error: error.message,
        stderr: stderr
      });
    }

    res.json({
      success: true,
      path: `/tmp/reports/${filename}.${format}`,
      stdout: stdout
    });
  });
});

/**
 * TAINT FLOW #2: Path Traversal
 * Source: req.params.fileId (from React ReportGenerator)
 * Sink: fs.readFile()
 * Path: URL parameter → file system read
 */
router.get('/files/:fileId/download', async (req: Request, res: Response) => {
  const { fileId } = req.params;

  // TAINT: User controlled path without sanitization
  // Attacker payload: "../../../etc/passwd" or "....//....//etc/passwd"
  const filePath = path.join('/var/uploads', fileId);

  // No validation that filePath is within allowed directory
  fs.readFile(filePath, (error, data) => {
    if (error) {
      return res.status(404).json({
        success: false,
        error: 'File not found',
        path: filePath  // Leaking full path
      });
    }

    res.setHeader('Content-Type', 'application/octet-stream');
    res.setHeader('Content-Disposition', `attachment; filename="${fileId}"`);
    res.send(data);
  });
});

/**
 * TAINT FLOW #3: SSRF (Server-Side Request Forgery)
 * Source: req.body.url (from React WebhookManager)
 * Sink: axios.get()
 * Path: HTTP POST body → server-side HTTP request
 */
router.post('/webhooks/register', async (req: Request, res: Response) => {
  const { url, events } = req.body;

  // TAINT: User controlled URL used for server-side request
  // Attacker payload: "http://169.254.169.254/latest/meta-data/" (AWS)
  // Or: "http://localhost:6379/CONFIG%20SET%20dir%20/tmp" (Redis)
  try {
    // Validating the webhook by making a request to it
    const response = await axios.get(url, {
      timeout: 5000,
      // No URL validation, allows internal network access
    });

    // Store webhook (simulated)
    const webhookId = `webhook_${Date.now()}`;

    res.json({
      success: true,
      id: webhookId,
      status: response.status,
      events: events
    });
  } catch (error: any) {
    res.status(400).json({
      success: false,
      error: error.message,
      // Leaking internal network information
      details: error.response?.data || 'Connection failed'
    });
  }
});

/**
 * TAINT FLOW #4: SQL Injection via Raw Query
 * Source: req.body.filter (from React advancedSearch)
 * Sink: sequelize.query() with raw SQL
 */
router.post('/users/advanced-search', async (req: Request, res: Response) => {
  const { filter } = req.body;

  // TAINT: User controlled filter object concatenated into SQL
  // Attacker payload: { "username": "admin'--" }
  const whereClause = Object.entries(filter)
    .map(([key, value]) => `${key} = '${value}'`)  // Direct string interpolation!
    .join(' AND ');

  const query = `SELECT * FROM users WHERE ${whereClause}`;

  try {
    const results = await sequelize.query(query, {
      type: QueryTypes.SELECT
    });

    res.json({
      success: true,
      data: results,
      query: query  // Leaking the SQL query
    });
  } catch (error: any) {
    res.status(500).json({
      success: false,
      error: error.message,
      query: query  // Leaking query even on error
    });
  }
});

/**
 * TAINT FLOW #5: Reflected XSS
 * Source: req.query.callback (JSONP callback)
 * Sink: Response body
 */
router.get('/api/jsonp', (req: Request, res: Response) => {
  const { callback, data } = req.query;

  // TAINT: User controlled callback name reflected in response
  // Attacker payload: callback=alert(document.cookie)//
  res.setHeader('Content-Type', 'application/javascript');
  res.send(`${callback}(${JSON.stringify({ data })})`);
});

/**
 * TAINT FLOW #6: Header Injection
 * Source: req.query.redirect (redirect URL)
 * Sink: res.setHeader('Location', ...)
 */
router.get('/redirect', (req: Request, res: Response) => {
  const { redirect } = req.query;

  // TAINT: User controlled redirect URL
  // Attacker payload: "http://evil.com" or with CRLF injection
  res.setHeader('Location', redirect as string);
  res.status(302).send();
});

/**
 * TAINT FLOW #7: Deserialization
 * Source: req.body.serialized (serialized object)
 * Sink: eval() or JSON.parse with reviver
 */
router.post('/deserialize', (req: Request, res: Response) => {
  const { serialized, type } = req.body;

  // TAINT: Deserializing user controlled data
  let result;
  if (type === 'json') {
    result = JSON.parse(serialized);
  } else if (type === 'eval') {
    // Extremely dangerous - code execution
    result = eval(`(${serialized})`);
  }

  res.json({ success: true, result });
});

/**
 * TAINT FLOW #8: Log Injection
 * Source: req.body.username (user input)
 * Sink: console.log() / file logger
 */
router.post('/login-attempt', (req: Request, res: Response) => {
  const { username, password } = req.body;

  // TAINT: User controlled data written to logs
  // Attacker payload: "admin\n[ERROR] System compromised\n[INFO] User: "
  console.log(`[INFO] Login attempt for user: ${username}`);

  // Simulate failed login
  res.status(401).json({ error: 'Invalid credentials' });
});

export default router;

File: frameworks/fastapi_project/taint_routes.py
"""
FastAPI Taint Flow Routes

These routes demonstrate cross-boundary taint flows from Vue frontend.
Each route has a clear Source → Sink path for taint analysis testing.

TAINT FLOWS:
1. /items/search - SQL Injection (query → SQL)
2. /upload/ - Path Traversal (file_path → open())
3. /items/import-xml - XXE (xml_data → lxml parser)
4. /items/{id}/email-preview - SSTI (template → Jinja2)
5. /eval - Code Injection (code → eval)
"""

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel
from typing import Optional, Any
import os
import subprocess
from xml.etree import ElementTree as ET

# Intentionally using vulnerable XML parser
try:
    from lxml import etree
    HAS_LXML = True
except ImportError:
    HAS_LXML = False

router = APIRouter()


class FileUpload(BaseModel):
    file_path: str
    content: str


class TemplateRequest(BaseModel):
    template: str


class EvalRequest(BaseModel):
    code: str
    context: Optional[dict] = None


class CommandRequest(BaseModel):
    command: str
    args: Optional[list] = None


# -----------------------------------------------------------------------------
# TAINT FLOW #1: SQL Injection
# Source: query parameter
# Sink: SQL query execution
# -----------------------------------------------------------------------------
@router.get("/items/search-vulnerable")
async def search_items_vulnerable(query: str, sort_by: str = "name"):
    """
    Vulnerable search endpoint with SQL injection.
    Vue frontend: ProductSearch.vue → productApi.searchProducts()
    """
    # TAINT: User input directly concatenated into SQL
    sql_query = f"SELECT * FROM items WHERE name LIKE '%{query}%' ORDER BY {sort_by}"

    return {
        "query": sql_query,  # Leaking the query
        "message": "Query would be executed (simulated)",
        "injection_point": f"query='{query}', sort_by='{sort_by}'"
    }


# -----------------------------------------------------------------------------
# TAINT FLOW #2: Path Traversal / Arbitrary File Write
# Source: file_path in request body
# Sink: open(file_path, 'w')
# -----------------------------------------------------------------------------
@router.post("/upload-vulnerable/")
async def upload_vulnerable(upload: FileUpload):
    """
    Vulnerable file upload with path traversal.
    Vue frontend: FileUploader.vue → productApi.uploadProductImage()
    """
    # TAINT: User controlled file path without sanitization
    # Attacker payload: "../../../etc/cron.d/malicious"
    file_path = upload.file_path

    # No validation that path is within allowed directory
    try:
        # This would write to arbitrary locations
        with open(file_path, 'w') as f:
            f.write(upload.content)

        return {
            "success": True,
            "message": f"File written to {file_path}",
            "path": file_path,
            "size": len(upload.content)
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "attempted_path": file_path
        }


# -----------------------------------------------------------------------------
# TAINT FLOW #3: XXE (XML External Entity)
# Source: XML body
# Sink: lxml.etree.parse() with resolve_entities=True
# -----------------------------------------------------------------------------
@router.post("/items/import-xml")
async def import_xml(request: Request):
    """
    Vulnerable XML import with XXE.
    Vue frontend: FileUploader.vue → productApi.importProductsXml()
    """
    xml_body = await request.body()

    # TAINT: Parsing XML with external entities enabled
    # Attacker payload:
    # <?xml version="1.0"?>
    # <!DOCTYPE foo [<!ENTITY xxe SYSTEM "file:///etc/passwd">]>
    # <products><name>&xxe;</name></products>

    if HAS_LXML:
        # Vulnerable: resolve_entities=True allows XXE
        parser = etree.XMLParser(resolve_entities=True, no_network=False)
        try:
            tree = etree.fromstring(xml_body, parser)
            return {
                "success": True,
                "parsed": etree.tostring(tree, encoding='unicode'),
                "warning": "XXE vulnerability - external entities resolved"
            }
        except Exception as e:
            return {"error": str(e)}
    else:
        # Fallback to stdlib (also vulnerable in older Python)
        try:
            tree = ET.fromstring(xml_body)
            return {
                "success": True,
                "parsed": ET.tostring(tree, encoding='unicode')
            }
        except Exception as e:
            return {"error": str(e)}


# -----------------------------------------------------------------------------
# TAINT FLOW #4: SSTI (Server-Side Template Injection)
# Source: template in request body
# Sink: Jinja2 Template().render()
# -----------------------------------------------------------------------------
@router.post("/items/{item_id}/email-preview")
async def email_preview(item_id: str, template_req: TemplateRequest):
    """
    Vulnerable template rendering with SSTI.
    Vue frontend: TemplateEditor.vue → productApi.renderProductEmail()
    """
    from jinja2 import Template

    # Simulated product data
    product = {
        "id": item_id,
        "name": "Test Product",
        "price": 29.99,
        "description": "A great product"
    }

    shop = {
        "name": "Anarchy Shop",
        "url": "http://localhost:8000"
    }

    # TAINT: User controlled template string rendered by Jinja2
    # Attacker payload: {{ config.items() }} or {{ self.__init__.__globals__ }}
    try:
        template = Template(template_req.template)
        rendered = template.render(product=product, shop=shop)

        return {
            "success": True,
            "rendered": rendered,
            "template_length": len(template_req.template)
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "template": template_req.template  # Leaking template on error
        }


# -----------------------------------------------------------------------------
# TAINT FLOW #5: Code Injection via eval()
# Source: code in request body
# Sink: eval()
# -----------------------------------------------------------------------------
@router.post("/eval")
async def dangerous_eval(eval_req: EvalRequest):
    """
    Extremely dangerous eval endpoint.
    """
    # TAINT: Directly executing user controlled code
    try:
        result = eval(eval_req.code, {"__builtins__": __builtins__}, eval_req.context or {})
        return {
            "success": True,
            "result": str(result),
            "code": eval_req.code
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "code": eval_req.code
        }


# -----------------------------------------------------------------------------
# TAINT FLOW #6: Command Injection
# Source: command in request body
# Sink: subprocess.run() with shell=True
# -----------------------------------------------------------------------------
@router.post("/system/exec")
async def exec_command(cmd_req: CommandRequest):
    """
    Vulnerable command execution endpoint.
    """
    # TAINT: User controlled command executed in shell
    # Attacker payload: "ls; cat /etc/passwd"
    command = cmd_req.command
    if cmd_req.args:
        command = f"{command} {' '.join(cmd_req.args)}"

    try:
        result = subprocess.run(
            command,
            shell=True,  # Vulnerable: shell=True allows command injection
            capture_output=True,
            text=True,
            timeout=30
        )

        return {
            "success": True,
            "stdout": result.stdout,
            "stderr": result.stderr,
            "returncode": result.returncode,
            "command": command
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "command": command
        }


# -----------------------------------------------------------------------------
# TAINT FLOW #7: SSRF
# Source: url in query parameter
# Sink: requests.get()
# -----------------------------------------------------------------------------
@router.get("/fetch-url")
async def fetch_url(url: str):
    """
    Vulnerable URL fetcher - SSRF.
    """
    import httpx

    # TAINT: User controlled URL for server-side request
    # Attacker payload: http://169.254.169.254/latest/meta-data/
    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(url, timeout=10.0)

        return {
            "success": True,
            "status_code": response.status_code,
            "content": response.text[:1000],  # First 1000 chars
            "url": url
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "url": url
        }


# -----------------------------------------------------------------------------
# TAINT FLOW #8: Arbitrary File Read
# Source: filename in path parameter
# Sink: open().read()
# -----------------------------------------------------------------------------
@router.get("/files/{filename:path}")
async def read_file(filename: str):
    """
    Vulnerable file read endpoint.
    """
    # TAINT: User controlled filename
    # Attacker payload: ../../etc/passwd
    try:
        with open(filename, 'r') as f:
            content = f.read()

        return {
            "success": True,
            "filename": filename,
            "content": content,
            "size": len(content)
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "filename": filename
        }
